de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/ProjectNegotiationCollector.java

```diff
@@ -0,0 +1,122 @@
+package de.fu_berlin.inf.dpp.negotiation;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Random;
+
+import org.apache.log4j.Logger;
+
+import de.fu_berlin.inf.dpp.filesystem.IProject;
+import de.fu_berlin.inf.dpp.filesystem.IResource;
+import de.fu_berlin.inf.dpp.session.ISarosSession;
+
+/**
+ * This class collects resources that should be added to a session to add them
+ * at once in preparation of a new single project negotiation. The reason is to
+ * prevent multiple concurrent running project negotiations per user.
+ * 
+ * This class allows to hold the reference to a thread that should handle this
+ * new project negotiation.
+ */
+public class ProjectNegotiationCollector {
+    private static final Logger log = Logger
+        .getLogger(ProjectNegotiationCollector.class);
+
+    private final Map<IProject, List<IResource>> mapping = new HashMap<IProject, List<IResource>>();
+
+    private Thread worker = null;
+
+    /**
+     * Add project resource mappings for the next project negotiation.
+     * 
+     * @param projectResourcesMapping
+     *            project resource mappings to add
+     */
+    public synchronized void add(
+        Map<IProject, List<IResource>> projectResourcesMapping) {
+        for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
+            .entrySet()) {
+            final IProject project = mapEntry.getKey();
+            final List<IResource> resources = mapEntry.getValue();
+
+            boolean alreadyFullShared = mapping.containsKey(project)
+                && mapping.get(project) == null;
+
+            /* full shared project */
+            if (resources == null || alreadyFullShared) {
+                mapping.put(project, null);
+                continue;
+            }
+
+            /* update partial shared project */
+            List<IResource> mappedResources = mapping.get(project);
+            if (mappedResources == null) {
+                mapping.put(project, new ArrayList<IResource>(resources));
+            } else {
+                mappedResources.addAll(resources);
+            }
+        }
+    }
+
+    /**
+     * Adds all added mappings to the current session, clears the map for next
+     * collection and returns a list of projects that should be handled by a
+     * project negotiation.
+     * 
+     * @return the list of projects to share
+     */
+    public synchronized List<IProject> getProjectsToShare(
```

Comment:
The name of this method does not reflect what it really does. Accessing the session should be done in the SessionManager and not here. The collector here should mainly merge resources that are required when for the next session negotiation.

Reply:
I moved the code back in the manager.